### PR TITLE
Featured method to change language of the app

### PR DESCRIPTION
**Pull request by samisii**

### CODE REVIEW

**Review by JuusoMRusanen — CHANGES_REQUESTED** [review]
Request changes

File: src/components/providers/authentication-provider.tsx [hunk 1]
```diff
@@ -5,7 +5,7 @@ import Keycloak from "keycloak-js";
 import { useCallback, useEffect } from "react";
 
 interface Props {
-  children: JSX.Element;
+  children: JSX.Element[];
```
Comment (petri-nykanen) [inline src/components/providers/authentication-provider.tsx:8]: What does this array notation mean or change? Not that I know, just curious. Just seeing JSX.Element, ReactElement and ReactNode used interchangably.

Comment (Jdallos) [reply src/components/providers/authentication-provider.tsx:8]: Good question, chatGpt provided me with an informative answer!

Comment (Jdallos) [inline src/components/providers/authentication-provider.tsx:8]: Is this change necessary?

File: src/components/screens/dashboard/dashboard-screen.tsx [hunk 2]
```diff
@@ -1,19 +1,22 @@
-import { useAtom } from "jotai";
-import { authAtom } from "../../../atoms/auth";
+import React from "react";
```
Comment (Jdallos) [inline src/components/screens/dashboard/dashboard-screen.tsx:1]: Avoid importing React as done here, rather use destructured imports e.g. Import { useEffect } from React

File: src/components/screens/dashboard/dashboard-screen.tsx [hunk 3]
```diff
@@ -1,19 +1,22 @@
...
+import { useLocale } from "../../../localization/LanguageSwitcher";
 
 /**
  * Dashboard screen component
- * 
  */
 const DashboardScreen = () => {
+  const { locale } = useLocale();
   const [auth] = useAtom(authAtom);
+  React.useEffect(() => {
```
Comment (Jdallos) [inline src/components/screens/dashboard/dashboard-screen.tsx:11]: Change the above import then can just use useEffect here

File: src/components/screens/dashboard/dashboard-screen.tsx [hunk 4]
```diff
@@ -1,19 +1,22 @@
...
+import strings from "../../../localization/strings";
+import { useLocale } from "../../../localization/LanguageSwitcher";
 
 /**
  * Dashboard screen component
- * 
  */
 const DashboardScreen = () => {
+  const { locale } = useLocale();
   const [auth] = useAtom(authAtom);
```
Comment (Jdallos) [inline src/components/screens/dashboard/dashboard-screen.tsx:10]: useAtom is not imported which is causing an error preventing the dashboard-screen rendering

File: src/localization/strings.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import LocalizedStrings, { type LocalizedStringsMethods } from "localized-strings";
```
Comment (Jdallos) [inline src/localization/strings.ts:1]: type is not needed here

File: src/localization/strings.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ * Type representing the shape of your translation strings
+ */
+interface TranslationStrings {
+  notYetImplemented: string
+}
+
+/**
+ * Interface describing localized strings
+ */
+export interface IStrings extends LocalizedStringsMethods, TranslationStrings {
```
Comment (Jdallos) [inline src/localization/strings.ts:15]: Again this seems a little different to the example in OSS which should be the guide to follow.
e.g.export interface IStrings extends LocalizedStringsMethods, TranslationStrings could be 
export interface Localized extends LocalizedStringsMethods

const strings: IStrings = new LocalizedStrings could be
const strings: Localized = new LocalizedStrings

So really just a difference in naming, but I'd recommend keeping the names the same as they are in other projects,

File: src/localization/en.json [hunk 7]
```diff
@@ -0,0 +1,4 @@
+{
+    "notYetImplemented": "not yet implemented"
+
```
Comment (Jdallos) [inline src/localization/en.json:3]:
```suggestion
```

File: src/localization/LanguageSwitcher.tsx [hunk 8]
```diff
@@ -0,0 +1,79 @@
+import { type ReactNode, createContext, useContext, useState } from "react";
```
Comment (Jdallos) [inline src/localization/LanguageSwitcher.tsx:1]: I've not come across a LanguageSwitcher before, is this actually needed I don't see anything like this in OSS

File: src/localization/fi.json [hunk 9]
```diff
@@ -0,0 +1,3 @@
+{
+    "notYetImplemented": "ei vielä toteutettu"
+  }
```
Comment (Jdallos) [inline src/localization/fi.json:3]:
```suggestion
}
```

File: src/types/index.ts [hunk 10]
```diff
@@ -0,0 +1,4 @@
+/**
+ * Type describing available languages
+ */
+export type Language = "fi" | "en";
```
Comment (Jdallos) [inline src/types/index.ts:4]: As far as I can tell this type is unused and so can be removed

File: tsconfig.json [hunk 11]
```diff
@@ -1,21 +1,26 @@
 {
   "compilerOptions": {
     "target": "ES2020",
-    "useDefineForClassFields": true,
     "lib": ["ES2020", "DOM", "DOM.Iterable"],
     "module": "ESNext",
     "skipLibCheck": true,
     "useUnknownInCatchVariables": true,
     "strict": true,
+    "esModuleInterop": true,
```
Comment (Jdallos) [inline tsconfig.json:9]: Lets discuss the reasons for these config changes and check if they are necessary, I'd be hesitant to make any changes to the config without good reason

**Review by Jdallos — CHANGES_REQUESTED** [review]
Some changes requested, happy to discuss if needed

File: src/components/providers/authentication-provider.tsx (hunk 1 above)
Comment (samisii) [reply src/components/providers/authentication-provider.tsx:8]: Juuso came across with this solution. There was an error for multiple children, and I coudn't resolve it.

File: src/localization/LanguageSwitcher.tsx (hunk 8 above)
Comment (samisii) [reply src/localization/LanguageSwitcher.tsx:1]: Didn't realize that we'll reference OSS on this project, sorry. But this was actually just a "project name" for the file, did think in the beginning that it's not a good name for file but somehow ended up keeping it... Will rename appropriately.

File: src/components/providers/authentication-provider.tsx (hunk 1 above)
Comment (JuusoMRusanen) [reply src/components/providers/authentication-provider.tsx:8]: The code started working after this change

Comment (Jdallos) [reply src/components/providers/authentication-provider.tsx:8]: I removed it locally and it doesn't seem to cause any problems, also wasn't it working without this previously and it doesn't look like any other changes have being made in this file?

File: src/localization/LanguageSwitcher.tsx (hunk 8 above)
Comment (Jdallos) [reply src/localization/LanguageSwitcher.tsx:1]: I could be mistaken but I believe this whole file can be removed and is not necessary, we just use the localization package to achieve what we need. We can check this together though

File: src/components/screens/dashboard/dashboard-screen.tsx [hunk 12]
```diff
@@ -1,19 +1,27 @@
...
- * 
  */
 const DashboardScreen = () => {
+  const [locale] = useAtom(localeAtom);
+  
   const [auth] = useAtom(authAtom);
 
+  useEffect(() => {
+    strings.setLanguage(locale);
+  }, [locale]);
```
Comment (Jdallos) [inline src/components/screens/dashboard/dashboard-screen.tsx:17]: This works but will become problematic if we need to do this in every component. See the OSS atoms/language.ts you could use this file to replace you r current localeAtom.ts and this will take care of setting the language without having to use the useEffect in this way.

File: src/components/screens/dashboard/dashboard-screen.tsx [hunk 13]
```diff
@@ -1,19 +1,27 @@
...
+
   return (
-    <div>
-      <div>This is where we would put our dashboard! IF WE HAD ONE!</div>
+    <>
+      <div>{strings.notYetImplemented}</div>
       <button type="button" onClick={ auth?.logout }>Log out</button>
```
Comment (Jdallos) [inline src/components/screens/dashboard/dashboard-screen.tsx:22]: Logout string should be localized

File: src/types/index.ts
Comment (Jdallos) [inline src/types/index.ts:1]: Empty file/ folder should be removed

File: tsconfig.json [hunk 14]
```diff
@@ -1,21 +1,25 @@
...
-    "allowImportingTsExtensions": true,
+    "moduleResolution": "node",
     "resolveJsonModule": true,
     "isolatedModules": true,
     "noEmit": true,
     "jsx": "react-jsx",
   },
-  "include": ["src"],
-  "references": [{ "path": "./tsconfig.node.json" }]
+  "include": ["src/**/*.ts", "src/**/*.tsx", "localization/**/*.ts"],
```
Comment (Jdallos) [inline tsconfig.json:18]: As said in the previous comment, I don't think these changes to the config are necessary, revert this file so that is matches what is currently in develop.

**Review by Jdallos — CHANGES_REQUESTED** [review]
Changes requested

File: src/App.tsx [hunk 15]
```diff
@@ -20,11 +22,14 @@ function App () {
 
   return (
     <div className="App">
-      <ErrorHandler>
-        <AuthenticationProvider>
-          <RouterProvider router={router} />
-        </AuthenticationProvider>
-      </ErrorHandler>
+      <Provider>
```
[the remark this answers is not in the document]
Comment (samisii) [reply src/App.tsx:25]: You're right, it didn't do anything. LocaleProvider confused me a little. Deleted now

File: src/components/screens/dashboard/dashboard-screen.tsx (hunk 12 above)
Comment (samisii) [reply src/components/screens/dashboard/dashboard-screen.tsx:17]: Modified dashboard to use languageAtom

File: src/App.tsx [hunk 16]
```diff
@@ -23,12 +23,14 @@ function App () {
 
   return (
     <div className="App">
-        <ErrorHandler>
+      <ThemeProvider theme={theme}>
+      <ErrorHandler>
```
Comment (joonas-ris) [inline src/App.tsx:27]: Fix indent

File: src/components/screens/dashboard/dashboard-screen.tsx [hunk 17]
```diff
@@ -1,24 +1,63 @@
...
+ */
+  const getPersons = async () => {
+    setIsLoading(true);
+    const fetchedPersons = await personsApi.listPersons({ active: true });
+    const loggedInPerson = fetchedPersons.filter(person => person.keycloakId === userProfile?.id);
+
+    if (loggedInPerson.length) {
+      try {
+          const fetchedPerson = await personsApi.listPersonTotalTime({personId: loggedInPerson[0].id});
+          setPersonTotalTime(fetchedPerson[0]);
```
Comment (joonas-ris) [inline src/components/screens/dashboard/dashboard-screen.tsx:35]: Fix indent

File: src/components/locale-toggle/LocaleToggle.tsx [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+import { languageAtom } from '../../atoms/languageAtom';
+import { Language } from '../../types';
+
+/**
+ * Function to change language of the app
+ */
+function LocaleToggle() {
+    const [language, setLanguage] = useAtom(languageAtom);
+
+    const handleLocaleChange = (newLanguage: string) => {
```
Comment (Jdallos) [inline src/components/locale-toggle/LocaleToggle.tsx:13]: Missing JSDoc comments

File: src/components/locale-toggle/LocaleToggle.tsx [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+        setLanguage(newLanguage as Language);
+        strings.setLanguage(newLanguage);
```
Comment (Jdallos) [inline src/components/locale-toggle/LocaleToggle.tsx:15]: I think this is already being taken care of within the languageAtom changes you have made, so this line strings.setLanguage(newLanguage);
Can be removed here

File: src/components/screens/dashboard/dashboard-screen.tsx [hunk 20]
```diff
@@ -17,6 +19,7 @@ function DashboardScreen () {
   const [isLoading, setIsLoading] = useState(false);
   const [personTotalTime, setPersonTotalTime] = useState<PersonTotalTime>();
   const { personsApi } = useApi();
+  const [_language] = useAtom(languageAtom);
```
Comment (Jdallos) [inline src/components/screens/dashboard/dashboard-screen.tsx:22]: It looks like this is unused and marked as so, so can this just be removed?

File: src/localization/en.json [hunk 21]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (Jdallos) [inline src/localization/en.json:1]: As the app grows we will want to structure this more, for example could logout and currentLocale strings be in a header object?

File: src/localization/fi.json [hunk 22]
```diff
@@ -0,0 +1,9 @@
+{
+  "notYetImplemented": "Ei vielä toteutettu",
```
Comment (Jdallos) [inline src/localization/fi.json:2]: Update structure as said in comment for en.json

File: src/localization/strings.ts [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+  fi: string;
+}
+
+/**
+* Translations interface
+*/
+interface Translations {
+  notYetImplemented: string;
+  currentLocaleLabel: string;
+  localization: Localizations;
```
Comment (Jdallos) [inline src/localization/strings.ts:19]: Typically the localization would be defined in the main interface Translations (which is typically called Localized) rather than as a separate interface.

File: src/localization/strings.ts [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+  logout: string;
+}
+
+/**
+* Localized strings
+*/
+export interface Localized extends LocalizedStringsMethods, Translations {}
```
Comment (Jdallos) [inline src/localization/strings.ts:26]: I don't think there is any advantage to doing it this way. Typically the interface you have called Translations would be called Localized and extend LocalizedStringsMethods, rather than doing it here as you have done. Let me know if there is a reason for doing it this way I have missed

File: tsconfig.json
Comment (Jdallos) [inline tsconfig.json:1]: If there are no changes to this file you can right click on the file in the vsCode source control section and discard all changes for it.

**Review by Jdallos — CHANGES_REQUESTED** [review]

File: src/components/providers/authentication-provider.tsx (hunk 1 above)
Comment (JuusoMRusanen) [reply src/components/providers/authentication-provider.tsx:8]: Indeed does work without this change. Reversed the change.

**Comment by JuusoMRusanen** [comment]
Made changes on behalf of Sami

File: src/components/screens/dashboard/dashboard-screen.tsx [hunk 25]
```diff
@@ -6,53 +6,58 @@ import { PersonTotalTime } from "../../../generated/client";
 import { errorAtom } from "../../../atoms/error";
 import LoaderWrapper from "../../generics/loader-wrapper";
 import { getHoursAndMinutes } from "../../../utils/time-utils";
+import strings from "../../../localization/strings";
 
 /**
  * Dashboard screen component
  */
-function DashboardScreen () {
+function DashboardScreen() {
```
Comment (Jdallos) [inline src/components/screens/dashboard/dashboard-screen.tsx:14]: Components should be declared using arrow function syntax.

**Review by Jdallos — CHANGES_REQUESTED** [review]

**Review by Jdallos — CHANGES_REQUESTED** [review]

File: package.json
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply package.json:1]: I removed all the unused dependencies and nothing broke.

File: src/App.tsx [hunk 26]
```diff
@@ -10,8 +11,7 @@ import { theme } from "./theme";
  * Application component
  *
  */
-function App () {
-
+function App() {
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/App.tsx:14]: fixed

File: src/components/localization-buttons/LocalizationButtons.tsx
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/localization-buttons/LocalizationButtons.tsx:1]: Renamed folder to layout-components.

File: src/components/localization-buttons/LocalizationButtons.tsx [hunk 27]
```diff
@@ -0,0 +1,47 @@
+import { useAtom } from "jotai";
+import strings from "../../localization/strings";
+import { Button } from "@mui/material";
+import { languageAtom } from "../../atoms/languageAtom";
+import { Language } from "../../types";
+
+/**
+ * LocalizationButtons component
+ */
+const LocalizationButtons: React.FC = () => {
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/localization-buttons/LocalizationButtons.tsx:10]: Removed React.FC

File: src/components/screens/dashboard/dashboard-screen.tsx (hunk 25 above)
Comment (JuusoMRusanen) [reply src/components/screens/dashboard/dashboard-screen.tsx:14]: Changed to arrow function

File: src/components/localization-buttons/LocalizationButtons.tsx [hunk 28]
```diff
@@ -0,0 +1,47 @@
...
+      ))}
+    </div>
+  );
+
+  return (
+    <div>
+      {renderLocalizationButtons()}
+      <p>
+        {strings.label.currentLocaleLabel} {language}
+      </p>
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/localization-buttons/LocalizationButtons.tsx]: Came up with a different solution: Changed Button component to ButtonToggle component, which shows which localization is selected by highlighting the corresponding button.

**Review by Jdallos — CHANGES_REQUESTED** [review]

File: src/components/layout-components/localization-buttons.tsx [hunk 29]
```diff
@@ -0,0 +1,49 @@
...
+    newLanguage: Language | null
+  ) => {
+    setLanguage(newLanguage as Language);
+  };
+
+  /**
+   * Renders localization buttons
+   */
+  const renderLocalizationButtons = () => (
+    <div>
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/layout-components/localization-buttons.tsx:27]: Nope. Removed it.

File: src/components/layout-components/localization-buttons.tsx [hunk 30]
```diff
@@ -0,0 +1,49 @@
...
+ * LocalizationButtons component
+ */
+const LocalizationButtons = () => {
+  const [language, setLanguage] = useAtom(languageAtom);
+
+  /**
+   * Method to handle locale change
+   */
+  const handleLocaleChange = (
+    event: React.MouseEvent<HTMLElement>,
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/layout-components/localization-buttons.tsx:17]: Good advice. Fixed.

File: src/components/screens/dashboard/dashboard-screen.tsx
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/screens/dashboard/dashboard-screen.tsx:1]: Localized 'Your balance' string and fixed localization toggle.

File: src/components/layout-components/localization-buttons.tsx [hunk 31]
```diff
@@ -13,34 +14,29 @@ const LocalizationButtons = () => {
   /**
    * Method to handle locale change
    */
-  const handleLocaleChange = (
-    event: React.MouseEvent<HTMLElement>,
-    newLanguage: Language | null
-  ) => {
+  const handleLocaleChange = (_event: MouseEvent<HTMLElement>, newLanguage: Language | null) => {
```
Comment (Jdallos) [inline src/components/layout-components/localization-buttons.tsx:17]: Is there a reason newLanguage is typed as Language | null?
If so, it might be a good idea to check if the newLanguage is null and handle appropriately rather than end up setting the localization to null.

File: src/components/layout-components/localization-buttons.tsx [hunk 32]
```diff
@@ -13,34 +14,29 @@ const LocalizationButtons = () => {
...
-    </div>
+    <ToggleButtonGroup
+      value={language}
+      exclusive
+      onChange={handleLocaleChange}
+      aria-label="localization"
+    >
+      {strings.getAvailableLanguages().map((availableLanguage, idx) => {
+        const value = strings.getString(`localization.${availableLanguage}`, language);
+        return (
```
Comment (Jdallos) [inline src/components/layout-components/localization-buttons.tsx:33]: Having a line break before a return value is a good general rule for code readability

**Review by Jdallos — APPROVED** [review]
Approved with a couple of comments

File: src/components/layout-components/localization-buttons.tsx (hunk 31 above)
Comment (JuusoMRusanen) [reply src/components/layout-components/localization-buttons.tsx:17]: Removed null and added null check.

File: src/components/layout-components/localization-buttons.tsx (hunk 32 above)
Comment (JuusoMRusanen) [reply src/components/layout-components/localization-buttons.tsx:33]: Added a line break.
